File: packages/picoipc/cpp/examples/consumer.cpp
#include "smh_q/ring.hpp"

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <thread>

int main(int argc, char** argv) {
  const std::string name = (argc > 1) ? argv[1] : "smh_q_demo";
  const int timeout_ms = (argc > 2) ? std::atoi(argv[2]) : 5000;

  smh_q::Ring::Config cfg;
  cfg.name = name;
  cfg.slot_count = 16;
  cfg.slot_size = 128;

  smh_q::Ring ring(cfg, false);
  std::printf("consumer: name=%s timeout_ms=%d\n", name.c_str(), timeout_ms);

  const auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(timeout_ms);
  int received = 0;

  while (std::chrono::steady_clock::now() < deadline) {
    auto msg = ring.try_consume();
    if (msg.has_value()) {
      std::string text(reinterpret_cast<const char*>(msg->data()), msg->size());
      std::printf("consumed: %s (read_seq=%u write_seq=%u)\n", text.c_str(), ring.read_seq(),
                  ring.write_seq());
      ++received;
      continue;
    }

    const auto remaining = std::chrono::duration_cast<std::chrono::milliseconds>(
        deadline - std::chrono::steady_clock::now());
    if (remaining.count() <= 0) break;

    if (!ring.wait_readable(remaining)) {
      break;
    }
  }

  std::printf("consumer done: received=%d\n", received);
  return received > 0 ? 0 : 1;
}

File: packages/picoipc/python/examples/consumer.py
#!/usr/bin/env python3
"""Consumer: reads messages, blocking on futex when idle."""

import argparse
import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from smh_q import Ring


def main() -> int:
    p = argparse.ArgumentParser(description="smh_q consumer")
    p.add_argument("name", nargs="?", default="smh_q_demo")
    p.add_argument("timeout_ms", nargs="?", type=int, default=5000)
    args = p.parse_args()

    ring = Ring(args.name, create=False)
    print(f"consumer: name={args.name} timeout_ms={args.timeout_ms}")

    deadline = time.monotonic() + args.timeout_ms / 1000.0
    received = 0

    while time.monotonic() < deadline:
        msg = ring.try_consume()
        if msg is not None:
            print(
                f"consumed: {msg.decode()} "
                f"(read_seq={ring.read_seq()} write_seq={ring.write_seq()})"
            )
            received += 1
            continue

        rem_ms = int((deadline - time.monotonic()) * 1000)
        if rem_ms <= 0:
            break
        if not ring.wait_readable(rem_ms):
            break

    print(f"consumer done: received={received}")
    return 0 if received > 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())
